layout_stride_zero_strides/layout_stride_zero_strides.md

```diff
@@ -0,0 +1,391 @@
+---
+title: "Let `layout_stride::mapping` with zero extent(s) accept zero strides"
+document: PXXXXR0
+date: 2025-10-22
+audience: LEWG
+author:
+  - name: Jacob Faibussowitsch
+  - name: Mark Hoemmen
+    email: <mhoemmen@nvidia.com>
+  - name: Christian Trott
+    email: <crtrott@sandia.gov>
+toc: true
+---
+
+# Author
+
+* Jacob Faibussowitsch (NVIDIA)
+
+* Mark Hoemmen (mhoemmen@nvidia.com) (NVIDIA)
+
+* Christian Trott (crtrott@sandia.gov) (Sandia National Laboratories)
+
+# Revision history
+
+* Revision 0 to be submitted 2025-??-??
+
+# Introduction
+
+We propose to change `layout_stride::mapping`'s constructors
+to permit strides to be zero if one or more extents are zero.
+For example, for extents (3, 5, 0, 11),
+this change would permit strides (1, 3, 0, 105).
+Currently, users would need to set the stride(s) corresponding
+to zero extents to an arbitrary positive value, such as 1.
+That would make the strides (1, 3, 1, 105) in this example.
+
+This change has two benefits.
+First, it conforms with NumPy's `ndarray` convention for specifying strides.
+This would let users construct a `layout_stride::mapping`
+directly from a NumPy array, or from any library that follows this convention.
+Second, it would let users convert `layout_left::mapping`
+or `layout_right::mapping` with size-zero extents to `layout_stride::mapping`.
+
+This change would relax preconditions of two existing
+`layout_stride::mapping` constructors.
+It would not change what C++ code is currently well-formed or ill-formed.
+The only way in which it could affect backwards compatibility is by
+introducing `layout_stride::mapping` instances that have zero strides.
+This may affect use of some libraries like the BLAS and LAPACK
+that forbid a zero stride,
+even if the corresponding matrix dimension is zero.
+On the other hand, `layout_left` and `layout_right` mappings
+can already have zero strides,
+so generic code that operates on the elements of an `mdspan`
+by calling the BLAS or LAPACK must already account for this.
+
+# Motivation
+
+## Interoperability between NumPy arrays and `mdspan`
+
+[NumPy](https://numpy.org/) is a Python package for scientific computing.
+Its [`ndarray` protocol](https://numpy.org/doc/2.3/reference/arrays.ndarray.html)
+defines an application binary interface for sharing multidimensional array views
+across libraries, programming languages, and hardware interfaces.
+
+Since `ndarray` objects represent multidimensional strided array views,
+it would be natural to translate them to C++ by using `mdspan` with `layout_stride`.
+Translation from C++ `mdspan` with `layout_stride` to Python `ndarray`
+works just fine, because Python permits arbitrary strides
+if the multidimensional index space has size zero.
+We show how to do this conversion from C++ to Python below
+by using the `pybind11` library.

```

Comment:
_Technically_ you are also supposed to have stride of 0 for 0-extents in numpy arrays. "Proper" conversion implementations will abide by this to be conformant with natively numpy-generated arrays. In practice though it doesn't make a lot of difference (since those arrays are empty).

All this is to say: if we need to make the argument for allowing 0 strides stronger, we should add a paragraph here explaining this instead of claiming we can convert C++ -> Numpy without modification.